builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,598 @@
+		)),
+	)
+}
+
+// getNomadClient provides
+// the client connection used by resources to interact with Nomad.
+func (r *Releaser) getNomadClient() (*nomadClient, error) {
+	// Get our client
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+	return &nomadClient{
+		NomadClient: client,
+	}, nil
+}
+
+func (r *Releaser) resourceJobCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	target *Deployment,
+	result *Release,
+	state *Resource_Job,
+	client *nomadClient,
+	st terminal.Status,
+) error {
+	// Set up clients
+	jobClient := client.NomadClient.Jobs()
+	deploymentClient := client.NomadClient.Deployments()
+
+	st.Update("Getting job...")
+	jobs, _, err := jobClient.PrefixList(target.Name)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad jobs: %s", err.Error())
+	}
+
+	if len(jobs) > 0 {
+		if target.Name != jobs[0].ID {
+			st.Step(terminal.StatusError, fmt.Sprintf("Job could not be found, did you mean to promote %q?", jobs[0].ID))
+			return nil
+		}
+	} else {
+		st.Step(terminal.StatusError, "Job not found.")
+		return nil
+	}
+
+	q := &api.QueryOptions{
+		Namespace: jobs[0].JobSummary.Namespace,
+	}
+
+	job, _, err := jobClient.Info(jobs[0].ID, q)
+
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad job: %s", err.Error())
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job.")
+		return nil
+	} else if deploy.JobVersion != *job.Version {
+		st.Update("Job version does not match deployment's job version.")
+		return nil
+	} else if deploy.Status != "running" {
+		st.Update("Deployment for job is no longer running.")
+		return nil
+	}
+
+	canaryDeployment := false
+	groupsToPromote := make([]string, len(deploy.TaskGroups))
+	for taskGroupName, taskGroup := range deploy.TaskGroups {
+		if r.config.Groups != nil {
+			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries > 0 {
+				canaryDeployment = true
+				groupsToPromote = append(groupsToPromote, taskGroupName)
+				continue
+			}
+		} else if taskGroup.DesiredCanaries > 0 {
+			canaryDeployment = true
+			// If no groups to promote are specified in the config, promote all groups
+			//   that have canaries
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
+	}
+	if !canaryDeployment {
+		st.Step(terminal.StatusWarn, "Nomad canary allocations not detected in job task groups.")
+		log.Info("Canaries not detected")
+		return nil
+	}
+
+	var currentTaskGroupState *api.DeploymentState
+	for _, group := range groupsToPromote {
+		if group != "" {
+			st.Update(fmt.Sprintf("Checking task group: %s", group))
```

Comment:
I'd love it if we could also print some health information here from Nomad allocation while we wait. Similar to what we do with the Kubernetes plugin:

```
» Deploying tetris...
✓ Kubernetes client connected to https://127.0.0.1:36367 with namespace default
✓ Creating deployment...
✓ Expected "http" port "3000" for app "tetris-v1"
⠏ Waiting on deployment to become available: requested=1 running=0 ready=0
```

Otherwise for unhealthy deployments, it is unclear what is being checked or why it hasn't progressed. I pasted some extra output from the Nomad CLI that I was looking at while waiting for the task group to come up. I think maybe we could show some of this info from `currentTaskGroupState` if possible.

```
Latest Deployment
ID          = 89389c34
Status      = running
Description = Deployment is running but requires manual promotion

Deployed
Task Group  Auto Revert  Promoted  Desired  Canaries  Placed  Healthy  Unhealthy  Progress Deadline
app         true         false     1        1         3       0        3          2022-02-24T14:18:15-08:00

Allocations
ID        Node ID   Task Group  Version  Desired  Status   Created     Modified
a4a534b0  51dfa101  app         1        run      failed   1m23s ago   43s ago
5da6f961  51dfa101  app         1        stop     failed   3m2s ago    1m23s ago
bfb30a7e  51dfa101  app         1        stop     failed   4m12s ago   3m2s ago
55861890  51dfa101  app         0        run      running  13m21s ago  13m11s ago

Time                       Type             Description
2022-02-24T14:10:03-08:00  Alloc Unhealthy  Unhealthy because of failed task
2022-02-24T14:10:03-08:00  Not Restarting   Exceeded allowed attempts 2 in interval 30m0s and mode is "fail"
2022-02-24T14:10:03-08:00  Terminated       Exit Code: 1, Exit Message: "Docker container exited with non-zero exit code: 1"
2022-02-24T14:10:02-08:00  Started          Task started by client
2022-02-24T14:09:45-08:00  Restarting       Task restarting in 16.755013793s
2022-02-24T14:09:45-08:00  Terminated       Exit Code: 1, Exit Message: "Docker container exited with non-zero exit code: 1"
2022-02-24T14:09:44-08:00  Started          Task started by client
2022-02-24T14:09:26-08:00  Restarting       Task restarting in 18.214921411s
2022-02-24T14:09:26-08:00  Terminated       Exit Code: 1, Exit Message: "Docker container exited with non-zero exit code: 1"
2022-02-24T14:09:24-08:00  Started          Task started by client
```

I haven't figured out why my Nomad deployment was unhealthy and the container kept exiting, it might be an issue on my end! But some UX I noticed could be improved :)

Reply:
If you have an idea of why it might be exiting let me know. I don't see any failures in docker logs or Nomad 🤔 I can provide Nomad jobspec and Waypoint HCL files of course if needed.

Reply:
@briancain the jobsepc and HCL would definitely help! I can take a look at that for you, & will look into how to improve the UX here.

Reply:
Since the pre-requisite for promoting a Nomad canary deployment is that all canary allocations are healthy, I added a line to display the # of healthy allocs vs. the number of deployed canaries. Once they're equal, that's when the promotion can happen!

```go
st.Update("Waiting on allocations to become healthy: healthy allocs=" + strconv.Itoa(currentTaskGroupState.HealthyAllocs) + " placed canaries=" + strconv.Itoa(len(currentTaskGroupState.PlacedCanaries)))
```

And the output looks like (couldn't copy it live from terminal so took screenshot):
![image](https://user-images.githubusercontent.com/83741749/155660167-0ed4393a-07a2-4300-8289-b17a55ccb8f2.png)

Reply:
Looks amazing! Love it - thanks for adding that in 👍🏻